consensus/hotstuff/committees/leader/leader_selection.go

```diff
@@ -155,39 +157,12 @@ func weightedRandomSelection(
 
 	leaders := make([]uint16, 0, count)
 	for i := 0; i < count; i++ {
-		// pick a random number from 0 (inclusive) to cumsum (exclusive). Or [0, cumsum)
-		randomness := rng.UintN(cumsum)
+		// pick a random number [1, cumsum]
+		randomness := rng.UintN(cumsum) + 1
 
 		// binary search to find the leader index by the random number
-		leader := binarySearchStrictlyBigger(randomness, weightSums)
-
+		leader, _ := slices.BinarySearch(weightSums, randomness)
```

Comment:
For correctness, we depend here on a subtle property
* we must find the element with the _smallest_ index that is equal or larger than the given value `randomness` 

Note that this property is **not** guaranteed by [`slices.BinarySearch`](https://pkg.go.dev/golang.org/x/exp/slices#BinarySearch) in the case of equality, according to the documentation. For example, consider the slice `weightSums = [0, 0, 2, 3, 3, 3, 8, 16, 16]` and assume we have `randomness = 3`. An index satisfying the API contract would be `3` but also `4` or `5`. However,  in our context `4` or `5` would be invalid.

Nevertheless, the [_implementation_](https://cs.opensource.google/go/x/exp/+/9ce248df:slices/sort.go;l=65-67)  satisfies our stronger requirements:
```golang
// search returns the leftmost position where f returns true, or len(x) if f
// returns false for all x. This is the insertion position for target in x,
// and could point to an element that's either == target or not.
```

In other words, we are relying on non-documented, internal implementation details here, which is not great. 
I see two options:
1. add detailed documentation here, explaining the additional requirement _and_ add solid test coverage to ensure that we would detect any future changes in `slices.BinarySearch` that violate our property. 
2. Keep the existing code: it already has solid test coverage and good documentation. There is no speed disadvantage compared  to `slices.BinarySearch` because it is literally the same implementation.

Reply:
This statement is not that convincing to me:
> As a side benefit new generic-based version is slightly faster

If you take a look at [the implementation](https://cs.opensource.google/go/x/exp/+/9ce248df:slices/sort.go;l=90-105), it is almost identical same. 2.2% speedup are marginal and my assumption is that you are seeing noise as opposed to a real speedup.

Reply:
The main goal of the diff to decrease the complexity.  I've verified that implementation does return the leftmost element, but given that we are considering that this property can change, then the only sustainable way is to maintain our own fork/reimplementation.